source/slang/slang-lower-to-ir.cpp

```diff
@@ -5275,34 +5300,95 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
         {
-            requirementKeys.add(getInterfaceRequirementKey(requirementDecl));
+            if (as<AssocTypeDecl>(member))
+            {
+                subContext->env->mapDeclToValue[member] = getBuilder()->getAssociatedType();
+            }
+        }
 
+        // First, compute the number of requirement entries that will be included in this
+        // interface type.
+        UInt operandCount = 0;
+        for (auto requirementDecl : decl->members)
+        {
+            operandCount++;
             // As a special case, any type constraints placed
             // on an associated type will *also* need to be turned
             // into requirement keys for this interface.
             if (auto associatedTypeDecl = as<AssocTypeDecl>(requirementDecl))
             {
-                for (auto constraintDecl : associatedTypeDecl->getMembersOfType<TypeConstraintDecl>())
-                {
-                    requirementKeys.add(getInterfaceRequirementKey(constraintDecl));
-                }
+                operandCount += associatedTypeDecl->getMembersOfType<TypeConstraintDecl>().getCount();
             }
         }
 
+        // Allocate an IRInterfaceType with the `operandCount` operands.
+        IRInterfaceType* irInterface = subBuilder->createInterfaceType(operandCount, nullptr);
 
-        NestedContext nestedContext(this);
-        auto subBuilder = nestedContext.getBuilder();
-        auto subContext = nestedContext.getContext();
+        // Add `irInterface` to decl mapping now to prevent cyclic lowering.
+        setValue(subContext, decl, LoweredValInfo::simple(irInterface));
 
-        // Emit any generics that should wrap the actual type.
-        emitOuterGenerics(subContext, decl, decl);
+        UInt entryIndex = 0;
 
-        IRInterfaceType* irInterface = subBuilder->createInterfaceType(
-            requirementKeys.getCount(),
-            reinterpret_cast<IRInst**>(requirementKeys.getBuffer()));
+        for (auto requirementDecl : decl->members)
+        {
+            auto entry = subBuilder->createInterfaceRequirementEntry(
+                getInterfaceRequirementKey(requirementDecl),
+                nullptr);
+            IRInst* requirementVal = lowerDecl(subContext, requirementDecl).val;
+            if (requirementVal)
+            {
+                auto reqType = requirementVal->getFullType();
+                entry->setRequirementVal(reqType);
+                if (!requirementVal->hasUses())
+                {
+                    // Remove lowered `IRFunc`s since we only care about
+                    // function types.
+                    switch (requirementVal->op)
+                    {
+                    case kIROp_Func:
+                    case kIROp_Generic:
+                        requirementVal->removeAndDeallocate();
+                        break;
+                    default:
+                        break;
+                    }
+                }   
+            }
```

Comment:
This seems fine for now, but it would be cleaner to not even emit the `requirementDecl` in the first place, and instead just emit the *type* of `requirementDecl`.

The biggest problem in doing that right now is that the front-end doesn't store `Type` information on every `Decl`, and instead it has this complicated logic in `getTypeForDeclRef` that is used to compute the type based on the class of declaration.

We could either introduce similar logic in a subroutine here (something like `lowerDeclType()`) that looks for the different cases of declaration (or uses a visitor) and emits the appropriate type, *or* we could try to be better about having all front-end `Decl`s have a type directly stored on them, if it would make things easier downstream.

Reply:
I thought about adding `lowerDeclType` but decided that it is adding too much complexity. It's likely that may turn out to be the better choice down the road, but for now I think we can get by with this less efficient hack.